extensions/sql/contract-definition/store/src/main/java/org/eclipse/dataspaceconnector/sql/contractdefinition/store/SqlContractDefinitionStore.java

```diff
@@ -196,19 +152,37 @@ public ContractDefinition deleteById(String id) {
 
     }
 
+    private <T> T fromJson(String json, Class<T> clazz) {
+        try {
+            return objectMapper.readValue(json, clazz);
+        } catch (JsonProcessingException e) {
+            throw new EdcException(e);
+        }
+    }
+
+    private String toJson(Object object) {
+        try {
+            return objectMapper.writeValueAsString(object);
+        } catch (JsonProcessingException e) {
+            throw new EdcException(e);
+        }
+    }
+
     private boolean existsById(String definitionId, Connection connection) {
-        return executeQuery(connection, (resultSet) -> resultSet.getLong(1), SQL_COUNT_CLAUSE_TEMPLATE, definitionId).iterator().next() <= 0;
+        return transactionContext.execute(() -> executeQuery(connection, (resultSet) -> resultSet.getLong(1), statements.getCountTemplate(), definitionId).iterator().next() <= 0);
     }
 
     private ContractDefinition findById(Connection connection, String id) {
-        var contractDefinitions = executeQuery(connection, this::mapResultSet, SQL_FIND_BY_CLAUSE_TEMPLATE, id);
-        if (contractDefinitions.isEmpty()) {
-            return null;
-        }
-        if (contractDefinitions.size() == 1) {
-            return contractDefinitions.get(0);
-        }
-        throw new IllegalStateException("Expected result set size of 1 but got " + contractDefinitions.size());
+        return transactionContext.execute(() -> {
+            var contractDefinitions = executeQuery(connection, this::mapResultSet, statements.getFindByTemplate(), id);
+            if (contractDefinitions.isEmpty()) {
+                return null;
+            }
+            if (contractDefinitions.size() == 1) {
+                return contractDefinitions.get(0);
+            }
+            throw new IllegalStateException("Expected result set size of 1 but got " + contractDefinitions.size());
+        });
```

Comment:
This can be executed outside from transaction and with just single line "return single(executeQuery(connection, this::mapResultSet, statement.getFindByTemplate(), id)" and of course by catching the SQLException.

Reply:
I agree on adding a `single` method, but the exception must be thrown inside the transaction in order to trigger a rollback.